src/OpenTelemetry/Logs/BatchLogRecordExportProcessor.cs

```diff
@@ -57,9 +57,14 @@ public override void OnEnd(LogRecord data)
             // happen here.
             Debug.Assert(data != null, "LogRecord was null.");
 
-            data!.BufferLogScopes();
+            data!.Buffer();
 
-            base.OnEnd(data);
+            data.AddReference();
+
+            if (!this.TryExport(data))
+            {
+                LogRecordSharedPool.Current.Return(data);
```

Comment:
I think we should decrement the reference count before calling returning the data to pool here.

Reply:
So at the moment `Return` decrements the reference count. If it is `!= 0` then it no-ops. If it is `== 0` then it returns. So if I move that logic out, anywhere `Return` is called has to do more work. Something like...

```csharp
if (logRecord.RemoveReference() == 0)
{
   LogRecordSharedPool.Current.Return(data);
}
```

There are currently 5 places where `Return` is called. 1 more for `LogEmitter` (not on this PR).

I don't love duplicating that logic all over the place. What if I renamed the method `Return` -> `TryReturn` to better indicate what is going on?

Reply:
I was actually referring to this particular situation where adding the data to the Circular buffer fails. I was still wrong though. I got a little confused as the `Return` here would always be a no-op (as reference count is 2) but I then realized that the `Return` in `OpenTelemetryLogger.Log` would work:

```csharp
if (!this.TryExport(data))
{
    LogRecordSharedPool.Current.Return(data);
}
```